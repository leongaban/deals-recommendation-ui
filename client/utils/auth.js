/**
 * Rights that are valid for access to this application.
 * @const {!Array.<string>}
 */
const VALID_RIGHTS = ['admin', 'viewer'];

/**
 * Determines if the user has the proper rights to access the application.
 * @param {!Array.<!Object>} services The services to check for authorization.
 * @return {boolean}
 */
export const hasRights = (services = []) => {
  if (!services) return false;

  const siskelRights = services.find(svc => svc.name.toLowerCase() === 'siskel' && !svc.disabled);

  // Abort if the user doesn't have anmy Mythor rights.
  if (!siskelRights || siskelRights.disabed) return false;

  // Determine if the user has valid rights (and that those rights are enabled).
  const { rights = [] } = siskelRights;
  return rights.some(({ disabled, name }) =>
    !disabled && VALID_RIGHTS.indexOf(name.toLowerCase()) !== -1);
};

/**
 * Validates whether or not the email meets prerequisites.
 * @param {string} email The value of the email.
 * @return {bool} Whether or not the email is valid.
 */
export const validateEmail = email => email.match(/\w{1,}\.?\w{0,}@(yum|pizzahut)\.com/g);

/**
 * Validates whether or not the password meets prerequisites.
 * @param {string} pass The value of the password.
 * @return {bool} Whether or not the password is valid.
 * @NOTE: Only prerequisite is password length must be >= 6 characters
 */
export const validatePassword = pass => pass.length >= 6;

/**
 * Validates whether or not the the new passwords match
 * @param {string} newPass The value of the password.
 * @param {string} newConfirmPass The value of the second password (confirm password).
 * @return {bool} Whether or not the given passwords are valid and are identical.
 * @NOTE: Only prerequisite is password length must be >= 6 characters and
 ** the new passwords are identical
 */
export const validateChangePassword = (newPass, newConfirmPass) =>
  newPass === newConfirmPass && newConfirmPass.length >= 6;
