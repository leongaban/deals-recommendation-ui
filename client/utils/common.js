/**
 * Checks if passed in item is an Object and is emtpy.
 * @param {*} a The value to test for Object-ness.
 * @return {boolean} Whether or not this is an Object (that is NOT empty).
 */
export const isEmptyObj = obj => JSON.stringify(obj) === '{}';

/**
 * Checks if passed in item is an Array and is emtpy.
 * @param {*} a The value to test for Array-ness.
 * @return {boolean} Whether or not this is an Array (that is NOT empty).
 */
export const isArrayEmpty = a => (!Array.isArray(a) || !a.length);

/**
 * Checks if passed in item is an empty String.
 * @param {*} a The value to test for String-ness.
 * @return {boolean} Whether or not this is an empty String.
 */
export const isEmptyString = a => (a === '');

/**
 * Checks if passed in item is Undefined.
 * @param {*} i The value test for Undefined-ness.
 * @return {boolean} Whether or not this is an Array (that is NOT empty).
 */
export const isUndefined = i => typeof i === 'undefined';
