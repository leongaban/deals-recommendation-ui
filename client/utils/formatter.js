import React from 'react';

/* eslint-disable no-bitwise */
export const capitalizeFirst = string => string.charAt(0).toUpperCase() + string.slice(1);

export const truncate = (text, count = 20) => {
  const trimmed = text.substring(0, count);
  return text.length > count ? `${trimmed}...` : text;
};

// Generates a new GUID for use in notificationsReducer
export const guidGenerator = () => {
  const s4 = () => (((1 + Math.random()) * 0x10000) | 0).toString(16).substring(1);
  return (`${s4() + s4()}-${s4()}-${s4()}-${s4()}-${s4()}${s4()}${s4()}`);
};

/**
 * Creates a custom type classname (CURRENTLY ONLY FOR MYTHOR-FORM-FIELDS / MYTHOR-FORM-FIELD-MODAL)
 * @param {string} type The type of the component.
 * @return {string} The string 'mythor-form-modal mythor-form-modal-bool'
 */
export const getCustomClass = type => [
  'mythor-form-field-modal',
  `mythor-form-field${type ? `-${type.toLowerCase()}` : ''}`
].join(' ');

/**
 * Transform object into an array of objects.
 * @param {Object} obj The object you want to turn into an array of objects.
 * @return {Array<Object>} The array of objects created from original object passed in.
 */
export const transformObj = (obj) => {
  const keyArr = Object.keys(obj);
  const values = keyArr.map(k => obj[k]);
  const finalObjArr = [];

  values.reduce((acc, cur, index) => {
    const key = keyArr[index];

    finalObjArr.push(`//pizzahut.com${cur.location}/${key}`);
    return null;
  });

  return finalObjArr;
};

/**
* Removes a certain piece from the string passed in
* @param {String} string The string you want to remove from.
* @param {String} path The string you want matched and removed from 'string' parameter.
* @return {String} The result string with the path string removed.
*/
export const removePathFromString = (string, path) => {
  const pathRegEx = new RegExp(`${path}`, 'g');
  return string.replace(pathRegEx, '');
};

/**
 * Creates an array of objects from an array of strings (Meant for Akamai Images).
 * @param {Array} arr The array you want to iterate with.
 * @return {Array<Object>} The array of objects created from the array passed in.
 */
export const formatToArrayObjects = arr => arr.map((item, index) => {
  const strippedFile = item.substring(item.indexOf('images/'), item.length);
  return Object.assign({}, {
    id: index,
    icon: `http:${item}`,
    path: `http:${item}`,
    alt: strippedFile,
    file: removePathFromString(strippedFile, 'images/')
  });
});

export const formatIdAsDisplayName = id =>
  `${id}`.replace(/((?:^|\.)[a-z]|[A-Z])/g, char => ` ${char[char.length - 1].toUpperCase()}`).trim().replace('ies', 'y');

  /**
   * Returns JSX that is determined by the length of the array passed in (children).
   * NOTE: This is for the `ButtonGroup` component to determine whether to render an extra
   * div to keep the `Clone` and `Archive` buttons together.
   * If there's only two elements, it just renders them b/c the parent container is using
   * flexbox to spread them apart.
   * @param {!Array} children The array you want to iterate with.
   * @return {!Array<JSX>} The array of JSX elements in a layout determined by array
   */
export const determineLayout = (children = []) => {
  if (children.length > 2) {
    return ([
      <div className="clone-archive" key="first-two-buttons">
        { children.slice(0, 2) }
      </div>,
      children.slice(2, children.length)
    ]);
  }
  return children;
};

export const sortTable = (array, key, alphabetical) => array.slice().sort((a, b) => {
  const A = a[key].toString().toLowerCase();
  const B = b[key].toString().toLowerCase();
  if (alphabetical) {
    if (A < B) return -1;
    if (A > B) return 1;
    return 0;
  }
  if (A > B) return -1;
  if (A < B) return 1;
  return 0;
});
