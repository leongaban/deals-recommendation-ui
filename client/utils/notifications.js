import React from 'react';

import Notification from 'components/Common/notification';

export const findNotification = (id, notifications) =>
  (id && notifications.length > 0 ? notifications.find(n => n.id === id) : null);

export const showNotification = notification => (
  (notification && notification.id) ?
    <Notification
      className={`notification ${notification.type === 'error' ? 'error' : 'success'}`}
      open={!!notification.id}
      message={notification.message}
    />
    : null
);
