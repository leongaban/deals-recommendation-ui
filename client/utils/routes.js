import pathToRegexp from 'path-to-regexp';
import routesConfig from 'routes';

import { PAGE_AUTH_REQS } from 'copy/Global/routes';

const { ANY } = PAGE_AUTH_REQS;

const propDescriptor = {
  enumerable: false,
  writable: false,
  configurable: false
};

export const PROP_KEY = '[[Key]]';
export const PROP_REGEXP = '[[RegExp]]';
export const PROP_PARAMS = '[[Parameters]]';

export const specialArrToObj = a => a.reduce((o, v) => ({ ...o, [v[PROP_KEY]]: v }), {});

/**
 * Determine filter function
 * @param {string} filter Auth requirement
 * @param {function} filter Function used to filter array
 * @return {function} Function that should be used to filter
 */
export const getFilterFunc = (filter = null) => {
  if (filter === null || filter === undefined) return () => true;
  if (typeof filter === 'function') return filter;
  return r => r.authReq === filter || r.authReq === ANY;
};

export const getSpecialRouteObj = (key, route) =>
  Object.defineProperties(route, {
    // Save the key for back referencing
    [PROP_KEY]: { ...propDescriptor, value: key },
    // Convert route's path property into a regular expression
    [PROP_REGEXP]: { ...propDescriptor, value: pathToRegexp(route.path || '') },
    // Extract each parameter's name from route's path property using RegExp
    [PROP_PARAMS]: { ...propDescriptor, value: (route.path || '').match(/(:\w+)/ig) }
  });

/**
 * Gets routes config, filters, and returns in various formats.
 * @param {string=} filter Optional filter for the routes.
 * @return {object} Object containing Keys Array, Vals Array, and Object format
 */
export const getRoutesConfig = (filter) => {
  const routes = Object
    .keys(routesConfig)
    .map(k => getSpecialRouteObj(k, { ...routesConfig[k] }))
    .filter(getFilterFunc(filter));

  return {
    vals: routes,
    keys: routes.map(r => r[PROP_KEY]),
    obj: specialArrToObj(routes)
  };
};

/**
 * Gets a flattened array of routes for use in the Router.
 * @param {string=} filter Optional filter for the routes.
 * @return {!Array.<!Object>} Array of route objects.
 */
export const getRoutesArray = filter => getRoutesConfig(filter).vals;

/**
 * Gets a route object by matching pathname with defined route path
 * @param {string} pathname The path used for testing the defined route path
 * @param {string=} filter Optional filter for the routes.
 * @return {!Array.<!Object>} Array of route objects.
 */
export const getRouteFromPathName = (pathname, filter) =>
  // Test each route to see if it's a match for the pathname provided
  getRoutesConfig(filter).vals.find(route => route['[[RegExp]]'].test(pathname));
