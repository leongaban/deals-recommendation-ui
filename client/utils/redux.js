/* eslint-disable import/prefer-default-export, no-console */
import { initialStates } from 'reducers';

import { DEPRECATED } from 'copy/Global/common';
import {
  REDUX_WARN_PROP_DUPE,
  REDUX_WARN_PROP_MISSING,
  REDUX_WARN_PROP_DEPRECATED
} from 'copy/Global/redux';

const aggregateStates = { ...initialStates, routing: { location: null } };

const defaultPropDef = {
  value: null,
  writable: false,
  configurable: true,
  enumerable: true
};

const valueOf = item => item;
// @TODO: DO NOT DELETE THIS COMMENT
// Needs to be refactored in the future to avoid causing
// the entire SPA to refresh, but prevent mutation of deep
// redux state objects outside of reducers.
// if (!item) return item;
// if (Array.isArray(item)) return item.map(v => valueOf(v));
// if (typeof item === 'object' && item !== null) {
//   return Object.keys(item).reduce((ret, key) => ({
//     ...ret,
//     [key]: valueOf(item[key])
//   }), {});
// }
// return item;

// Grabs all appropriate initialStates to use in a container
const getters = Object
  .keys(aggregateStates)
  .reduce((gets, reducer) => (
    Object.keys(aggregateStates[reducer])
      .reduce((_, propName) => {
        if (propName in gets) {
          console.warn(REDUX_WARN_PROP_DUPE(propName, reducer));
          return gets;
        }

        return Object.defineProperty(gets, propName, {
          ...defaultPropDef,
          value: (state) => {
            if (aggregateStates[reducer][propName] === DEPRECATED) {
              console.warn(REDUX_WARN_PROP_DEPRECATED(propName, reducer));
            }
            return valueOf(state[reducer][propName]);
          }
        });
      }, gets)
  ), {});

/* eslint-disable key-spacing, no-multi-spaces */
export const cleanMapStateToProps = (propMap = []) => (state, props) => {
  let keyMap = [];
  let valMap = [];

  if (Array.isArray(propMap)) {
    keyMap = keyMap.concat(propMap);
    valMap = valMap.concat(propMap);
  } else if (typeof propMap === 'object') {
    keyMap = keyMap.concat(Object.keys(propMap));
    valMap = valMap.concat(keyMap.map(k => propMap[k]));
  } else {
    throw new TypeError(`cleanMapStateToProps( propMap ): Expected 'propMap' to be Object or Array, instead got ${typeof propMap}`);
  }

  return {
    ...props,
    ...keyMap.reduce((obj, key, i) => {
      const valKey = `${valMap[i]}`;
      if (valKey in getters) return { ...obj, [key]: getters[valKey](state, props) };
      if (valKey.indexOf(':') > -1) {
        const { entityReducer } = state;
        if (valKey.indexOf(':*') > -1) return { ...obj, ...valueOf(entityReducer) };
        if (valKey in entityReducer) return { ...obj, [key]: valueOf(entityReducer[valKey]) };
        return obj;
      }
      console.warn(REDUX_WARN_PROP_MISSING(valKey));
      return obj;
    }, {})
  };
};
