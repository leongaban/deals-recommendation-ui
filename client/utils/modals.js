// Utils
import { isUndefined } from 'utils/common';
import { capitalizeFirst as cap } from 'utils/formatter';

// Copy
import { PRODUCT_NAME } from 'copy/Components/modals';

// ProductModal
export const modalName = isEdit => (isEdit ? 'Edit' : 'Add');
export const enabledValue = value => (isUndefined(value) ? 'false' : value.toString());

// ProductModal (Edit)
export const setChecked = (type) => {
  switch (type) {
    case 'E': return 'Exclude';
    case 'O': return 'Override';
    case 'D':
    default: return 'Default';
  }
};

// ProductInput utils
const isDesc = id => id === 'description';
const renderDesc = edit => (edit ? cap(PRODUCT_NAME) : `Friendly ${PRODUCT_NAME}`);
export const classer = id => (isDesc(id) ? '' : 'col');
export const titler = (id, edit) => (isDesc(id) ? renderDesc(edit) : cap(id));
export const placeholder = id => (isDesc(id) ? 'Ex: Cookies' : '');
