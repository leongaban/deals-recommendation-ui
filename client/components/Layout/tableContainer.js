/* eslint-disable object-curly-newline */
import React, { Component } from 'react';
import { connect } from 'react-redux';

// Actions
import { getRules as fetchRules } from 'actions/Rules';

// Material-UI
import { withStyles } from '@material-ui/core/styles';
import AppBar from '@material-ui/core/AppBar';
import Tabs from '@material-ui/core/Tabs';
import Tab from '@material-ui/core/Tab';

// Utils
import { cleanMapStateToProps } from 'utils/redux';
import { isArrayEmpty } from 'utils/common';

// Components
import ProductsTabContainer from 'components/Layout/productsTabContainer';

// Constants
import { TABS_DEFINITIONS } from 'constants/tabs';

// Styles
import { styles } from './tableContainer.styles';

class TableContainer extends Component {
  constructor(props) {
    super(props);

    const {
      getRules,
      rulesAll
    } = props;

    this.state = { value: 0 };

    if (isArrayEmpty(rulesAll)) {
      getRules();
    }

    this.handleChangeTab = this.handleChangeTab.bind(this);
  }

  handleChangeTab(event, value) {
    this.setState({ value });
  }

  render() {
    const { classes } = this.props;
    const { value } = this.state;

    const { tabRoot, tabsRoot, tabSelected, tabsIndicator } = classes;
    const tabsClasses = { root: tabsRoot, indicator: tabsIndicator };
    const tabClasses = { root: tabRoot, selected: tabSelected };

    return (
      <div className={classes.root}>
        <AppBar position="static">
          <Tabs value={value} classes={tabsClasses} onChange={this.handleChangeTab}>
            {TABS_DEFINITIONS.map(({ label }) => (
              <Tab key={label} classes={tabClasses} label={label} />
            ))}
          </Tabs>
        </AppBar>
        <ProductsTabContainer value={value} {...this.props} />
      </div>
    );
  }
}

export const mapDispatchToProps = dispatch => ({
  getRules: (...args) => dispatch(fetchRules(...args))
});

const mapStateToProps = cleanMapStateToProps([
  'rulesAll',
  'rulesOverride',
  'rulesDefault',
  'rulesExcluded'
]);

export const TableContainerJest = TableContainer;

export default connect(mapStateToProps, mapDispatchToProps)(withStyles(styles)(TableContainer));
