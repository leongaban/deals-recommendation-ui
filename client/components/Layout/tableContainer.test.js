import { testCommonComponentAttrs, testDispatchToProps } from 'utils/tests';

import { TableContainerJest, mapDispatchToProps } from './tableContainer';

const mockGetRulesList = jest.fn(() => Promise.resolve());

jest.mock('actions/Rules', () => ({
  getRules: () => mockGetRulesList()
}));

const mockProps = {
  classes: {
    root: 'root'
  },
  getRules: mockGetRulesList,
  rulesAll: [],
  rulesOverride: [],
  rulesDefault: [],
  rulesExcluded: []
};

describe('<TableContainerJest /> component', () => {
  testCommonComponentAttrs(TableContainerJest, mockProps);
  testDispatchToProps(TableContainerJest, mapDispatchToProps);
});
