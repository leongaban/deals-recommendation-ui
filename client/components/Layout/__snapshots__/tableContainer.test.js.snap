// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TableContainerJest /> component when rendering should render a component matching the snapshot 1`] = `
<div
  className="root"
>
  <WithStyles(AppBar)
    position="static"
  >
    <WithStyles(Tabs)
      classes={
        Object {
          "indicator": undefined,
          "root": undefined,
        }
      }
      onChange={[Function]}
      value={0}
    >
      <WithStyles(Tab)
        classes={
          Object {
            "root": undefined,
            "selected": undefined,
          }
        }
        key="All Products"
        label="All Products"
      />
      <WithStyles(Tab)
        classes={
          Object {
            "root": undefined,
            "selected": undefined,
          }
        }
        key="Override"
        label="Override"
      />
      <WithStyles(Tab)
        classes={
          Object {
            "root": undefined,
            "selected": undefined,
          }
        }
        key="Default"
        label="Default"
      />
      <WithStyles(Tab)
        classes={
          Object {
            "root": undefined,
            "selected": undefined,
          }
        }
        key="Excluded"
        label="Excluded"
      />
    </WithStyles(Tabs)>
  </WithStyles(AppBar)>
  <ProductsTabContainer
    classes={
      Object {
        "root": "root",
      }
    }
    getRules={
      [MockFunction] {
        "calls": Array [
          Array [],
        ],
      }
    }
    rulesAll={Array []}
    rulesDefault={Array []}
    rulesExcluded={Array []}
    rulesOverride={Array []}
    value={0}
  />
</div>
`;
